engine/dlib/src/dlib/dstrings.cpp

```diff
@@ -231,3 +232,97 @@ int dmStrCaseCmp(const char *s1, const char *s2)
 #endif
 }
 
+#if defined(ANDROID)
+    #if defined(__USE_GNU) && __ANDROID_API__ >= 23
+        #define DM_STRERROR_USE_GNU
+    #else
+        #define DM_STRERROR_USE_POSIX
+    #endif
+#elif defined(__EMSCRIPTEN__)
+    // Emscripten wraps strerror_r as strerror anyway
+    #define DM_STRERROR_USE_UNSAFE
+#else
+    #if (_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600) && ! _GNU_SOURCE || defined(_WIN32) || defined(__MACH__)
+        #define DM_STRERROR_USE_POSIX
+    #else
+        #define DM_STRERROR_USE_GNU
+    #endif
+#endif
+
+#ifdef DM_STRERROR_USE_POSIX
+    #if defined(_WIN32)
+        #define DM_STRERROR_FN(buf, size, errval) (int) strerror_s(buf, size, errval)
+    #else
+        #define DM_STRERROR_FN(buf, size, errval) strerror_r(errval, buf, size)
+    #endif
+#endif
+
+void dmStrerror(char* dst, size_t size, int err)
+{
+    if (dst == 0 || size == 0)
+    {
+        return;
+    }
+
+    char scratch[256];
+    const size_t scratch_size = sizeof(scratch);
+    size_t err_msg_len = 0;
+    int old_errno = errno;
+    char* retstr = 0;
+
+#ifdef DM_STRERROR_USE_POSIX
+    int ret = DM_STRERROR_FN(scratch, scratch_size, err);
+
+    if (ret == 0 || ret == ERANGE)
+    {
+    #if defined(_WIN32)
+        // apparently windows returns a success ret value and a "Unknown error" string
+        // even if it's not a supported errno
+        if (dmStrCaseCmp("Unknown error", scratch) == 0)
+        {
+            dmSnPrintf(scratch, scratch_size, "Unknown error %d", err);
+        }
+    #endif
+    }
+    else
+    {
+        if (ret == EINVAL)
+        {
+            dmSnPrintf(scratch, scratch_size, "Unknown error %d", err);
+        }
+        else
+        {
+            dmSnPrintf(scratch, scratch_size, "Failed getting error (code %d)", ret);
+        }
+    }
+
+    retstr = &scratch[0];
+#endif
+
+#ifdef DM_STRERROR_USE_GNU
```

Comment:
Perhaps these could use "#elif defined(...", in order to ensure they aren't accidentally enabled at the same time?